### PR TITLE
Unit tests for FGS

### PR DESCRIPTION
**Pull request by smarella**
Added unit tests for testing the main functionality of the FGS classes and methods. The idea is to have a set of test classes that can be expanded as the functionality of FGS evolves/changes.

### CODE REVIEW

**Comment by smarella** [comment]
Rebased with phase1.

**Comment by smarella** [comment]
Rebased again with phase1. Fixed unit tests that were broken due to NM+Executor merge.

@kensipe Could you please take a look at this?

**Comment by kensipe** [comment]
@smarella I will look this week... sorry for delay

**Comment by kensipe** [comment]
I'll make comments against the test files... it in generally looks good.  there are things which would be idiomatically more groovy.  Also... for tests I tend to be more liberal with dynamic types... which is more a philosophical difference... I will make a comment but feel free to ignore. I will try to make it clear when it is philosophical.

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/SchedulerUtilsSpec.groovy [hunk 1]
```diff
@@ -29,6 +30,30 @@ class SchedulerUtilsSpec extends Specification {
 
     }
 
+    def "is eligible for FGS"() {
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/SchedulerUtilsSpec.groovy:33]: FGS is too terse IMO.  it makes sense now but 1 year from now no one will know what this is testing.

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/SchedulerUtilsSpec.groovy [hunk 2]
```diff
@@ -29,6 +30,30 @@ class SchedulerUtilsSpec extends Specification {
...
+        given:
+        def state = Mock(SchedulerState)
+        def tasks = new ArrayList<NodeTask>()
+        def fgsNMTask = new NodeTask(new NMProfile("zero", 0, 0))
+        def cgsNMTask = new NodeTask(new NMProfile("low", 2, 4096))
+        fgsNMTask.setHostname("test_fgs_hostname");
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/SchedulerUtilsSpec.groovy:39]: groovy doesn't need semi colons :)  also it is more idiomatic to do `fgsNMTask.hostname ="test_fgs_hostname"`

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/SchedulerUtilsSpec.groovy [hunk 3]
```diff
@@ -29,6 +30,30 @@ class SchedulerUtilsSpec extends Specification {
...
+        cgsNMTask.setHostname("test_cgs_hostname");
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/SchedulerUtilsSpec.groovy:40]: same as above

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/SchedulerUtilsSpec.groovy [hunk 4]
```diff
@@ -29,6 +30,30 @@ class SchedulerUtilsSpec extends Specification {
...
+        tasks.add(fgsNMTask)
+        tasks.add(cgsNMTask)
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/SchedulerUtilsSpec.groovy:42]: line 41 and 42 I would replace with `tasks << fgsNMTask << cgsNMTask`

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/SchedulerUtilsSpec.groovy [hunk 5]
```diff
@@ -29,6 +30,30 @@ class SchedulerUtilsSpec extends Specification {
...
+        state.getActiveTasks() >> tasks;
+
+        expect:
+        returnValue == SchedulerUtils.isEligibleForFineGrainedScaling(hostName, state);
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/SchedulerUtilsSpec.groovy:46]: don't need the semi-colon

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/SchedulerUtilsSpec.groovy [hunk 6]
```diff
@@ -29,6 +30,30 @@ class SchedulerUtilsSpec extends Specification {
 
     }
 
+    def "is eligible for FGS"() {
+        given:
+        def state = Mock(SchedulerState)
+        def tasks = new ArrayList<NodeTask>()
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/SchedulerUtilsSpec.groovy:36]: I would write: `def tasks = []`

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/FGSTestBaseSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    fgsContainer.container = Container.newInstance(containerId, node.getNodeID(), node.getHttpAddress(),
+        Resources.createResource(mem, cpu), null, null)
+    fgsContainer.rmContainer = new RMContainerImpl(fgsContainer.container, containerId.getApplicationAttemptId(),
+        node.getNodeID(), "user1", rmContext)
+    nodeStore.getNode(node.getNodeID().getHost()).getNode().allocateContainer(fgsContainer.rmContainer)
+    fgsContainer.containerStatus = ContainerStatus.newInstance(containerId, state, "", 0);
+    return fgsContainer
+  }
+
+}
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/FGSTestBaseSpec.groovy:152]: This class scares me... I can't even tell if there is a test in this... It has docs that indicate that it is a base class.   That would likely mean that we are extended for functions which is not good design to me.  I would favor utility classes and helper classes over a base class.  Also the name is too terse.  FGS will not mean anything in 1 year IMO

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/FGSTestBaseSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+/**
+ *
+ * Base class for FGS tests.
+ *
+ */
+class FGSTestBaseSpec extends Specification {
+  NodeStore nodeStore = new NodeStore();
+  SchedulerDriver mesosDriver = Mock(SchedulerDriver);
+  MyriadDriver myriadDriver = new MyriadDriver(mesosDriver);
+  OfferLifecycleManager offerLifecycleManager = new OfferLifecycleManager(nodeStore, myriadDriver);
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/FGSTestBaseSpec.groovy:36]: I would go with

```
  def nodeStore = new NodeStore();
  def mesosDriver = Mock(SchedulerDriver);
  def myriadDriver = new MyriadDriver(mesosDriver);
  def offerLifecycleManager = new OfferLifecycleManager(nodeStore, myriadDriver);

  def cfg = new MyriadConfiguration();
```

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/FGSTestBaseSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    ContainerId containerId = ContainerId.newContainerId(ApplicationAttemptId.newInstance(
+        ApplicationId.newInstance(123456789, 1), 1), cid)
+    FGSContainer fgsContainer = new FGSContainer()
+    fgsContainer.containerId = containerId
+    fgsContainer.container = Container.newInstance(containerId, node.getNodeID(), node.getHttpAddress(),
+        Resources.createResource(mem, cpu), null, null)
+    fgsContainer.rmContainer = new RMContainerImpl(fgsContainer.container, containerId.getApplicationAttemptId(),
+        node.getNodeID(), "user1", rmContext)
+    nodeStore.getNode(node.getNodeID().getHost()).getNode().allocateContainer(fgsContainer.rmContainer)
+    fgsContainer.containerStatus = ContainerStatus.newInstance(containerId, state, "", 0);
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/FGSTestBaseSpec.groovy:148]: semi-colon not necessary

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/FGSTestBaseSpec.groovy [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+        Thread.currentThread().getContextClassLoader().getResource("myriad-config-default.yml"),
+        MyriadConfiguration.class);
+  }
+/******************* Nodes Related ****************/
+
+  Map<NodeId, RMNode> rmNodes = new ConcurrentHashMap<NodeId, RMNode>();
+
+  RMNode getRMNode(int cpu, int mem, String host, Protos.SlaveID slaveId) {
+    RMNode rmNode = MockNodes.newNodeInfo(0, Resources.createResource(mem, cpu), 0, host)
+    if (rmNodes.containsKey(rmNode.getNodeID())) {
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/FGSTestBaseSpec.groovy:52]: groovy way:  `if (rmNodes[rmNode.getNodeID()]) {`

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/FGSTestBaseSpec.groovy [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+      throw new IllegalArgumentException("Node with hostname: " + host + " already exists")
+    }
+    rmNodes.put(rmNode.getNodeID(), rmNode)
+    nodeStore.add(getSchedulerNode(rmNode))
+    def node = nodeStore.getNode(host)
+    node.setSlaveId(slaveId)
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/FGSTestBaseSpec.groovy:58]: groovy: `node.slaveId = slaveId`

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/FGSTestBaseSpec.groovy [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+      }
+    }
+    return schedulerNode
+  }
+
+  /******************* RMContext Related ****************/
+
+  SystemMetricsPublisher publisher = Mock(SystemMetricsPublisher) {}
+  RMApplicationHistoryWriter writer = Mock(RMApplicationHistoryWriter) {}
+  EventHandler handler = Mock(EventHandler) {}
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/FGSTestBaseSpec.groovy:81]: probably incline to do:

```
  def publisher = Mock(SystemMetricsPublisher) {}
  def writer = Mock(RMApplicationHistoryWriter) {}
  def handler = Mock(EventHandler) {}
```

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/NMHeartBeatHandlerSpec.groovy [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    hbHandler.beforeRMNodeEventHandled(getNMRegistrationEvent(nonZeroNM), rmContext);
+
+    then:
+    1 * hbHandler.logger.warn('FineGrainedScaling feature got invoked for a NM with non-zero capacity. ' +
+        'Host: {}, Mem: {}, CPU: {}. Setting the NM\'s capacity to (0G,0CPU)', 'test_host1', 2048, 2)
+    nonZeroNM.getTotalCapability().getMemory() == 0
+    nonZeroNM.getTotalCapability().getVirtualCores() == 0
+
+    when:
+    hbHandler.beforeRMNodeEventHandled(getNMRegistrationEvent(zeroNM), rmContext);
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/NMHeartBeatHandlerSpec.groovy:39]: semi-colon not necessary :)

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/NMHeartBeatHandlerSpec.groovy [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+    then:
+    0 * hbHandler.logger.warn(_) // no logger.warn invoked
+    nonZeroNM.getTotalCapability().getMemory() == 0
+    nonZeroNM.getTotalCapability().getVirtualCores() == 0
+  }
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/NMHeartBeatHandlerSpec.groovy:45]: nice test

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/NMHeartBeatHandlerSpec.groovy [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+  NMHeartBeatHandler getNMHeartBeatHandler(YarnNodeCapacityManager yarnNodeCapacityMgr) {
+    def registry = Mock(InterceptorRegistry)
+    def state = Mock(SchedulerState)
+    return new NMHeartBeatHandler(registry, yarnScheduler, myriadDriver,
+        yarnNodeCapacityMgr, offerLifecycleManager, nodeStore, state)
+  }
+
+}
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/NMHeartBeatHandlerSpec.groovy:96]: philosophical:  I would have probably `def` the returns on these last 4 methods... so mine would have looked like:

```
  def getNMRegistrationEvent(RMNode node) {
    new RMNodeStartedEvent(node.getNodeID(), null, null)
  }
```

dynamic types can be challenge for some... that challenge is mostly based on lack on context.  the beauty of a test is that it is self contained and the context is right there... most of the time (unless you have base classes and utility class... I would be more expressive there)

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+ *
+ */
+class YarnNodeCapacityManagerSpec extends FGSTestBaseSpec {
+
+  def "No Containers Allocated Due To Mesos Offers"() {
+    given:
+    def yarnNodeCapacityMgr = getYarnNodeCapacityManager()
+
+    def host = "test_host"
+    def slaveId = Protos.SlaveID.newBuilder().setValue(host + "_slave_id").build()
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy:25]: a common groovy way to do this is:
`def slaveId = Protos.SlaveID.newBuilder().setValue("${host}_slave_id").build()`

however in this case yours may be more readable.

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+    // Node's capacity set to match the size of 2 containers + mesos offers
+    yarnNodeCapacityMgr.setNodeCapacity(zeroNM, Resources.createResource(6144, 6))
+
+    // no new container allocations
+
+    when:
+    yarnNodeCapacityMgr.handleContainerAllocation(zeroNM)
+
+    then:
+    nodeStore.getNode(host).getNode().getRunningContainers().size() == 2  // 2 containers still running
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy:46]: it is common to be more terse in groovy with chaining of properties... like this:

`nodeStore.getNode(host).node.runningContainers.size == 2  // 2 containers still running`

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+    1 * mesosDriver.declineOffer(offer.getId())   // offer rejected, as it's not used to allocate more containers
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy:47]: groovy: `1 * mesosDriver.declineOffer(offer.id)`

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+    zeroNM.getTotalCapability().getVirtualCores() == 2 // capacity returns back to match size of running containers
+    zeroNM.getTotalCapability().getMemory() == 2048
+    nodeStore.getNode(host).getContainerSnapshot() == null // container snapshot is released
+  }
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy:51]: move groovy:

```
    nodeStore.getNode(host).node.runningContainers.size == 2  // 2 containers still running
    1 * mesosDriver.declineOffer(offer.id)   // offer rejected, as it's not used to allocate more containers
    zeroNM.totalCapability.virtualCores == 2 // capacity returns back to match size of running containers
    zeroNM.totalCapability.memory == 2048
    nodeStore.getNode(host).containerSnapshot == null // container snapshot is released
```

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+    def fgsContainer1 = getFGSContainer(zeroNM, 1, 1, 1024, ContainerState.RUNNING)
+    def fgsContainer2 = getFGSContainer(zeroNM, 2, 1, 1024, ContainerState.RUNNING)
+    nodeStore.getNode(host).snapshotRunningContainers()
+
+    // Node's capacity set to match the size of 2 running containers + mesos offers
+    yarnNodeCapacityMgr.setNodeCapacity(zeroNM, Resources.createResource(6144, 6))
+
+    // 2 new containers allocated after HB
+    def fgsContainer3 = getFGSContainer(zeroNM, 3, 1, 1024, ContainerState.NEW)
+    def fgsContainer4 = getFGSContainer(zeroNM, 4, 1, 1024, ContainerState.NEW)
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy:75]: fgsContainer 1..4 are never used... why are they declared?

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+    when:
+    yarnNodeCapacityMgr.handleContainerAllocation(zeroNM)
+
+    then:
+    nodeStore.getNode(host).getNode().getRunningContainers().size() == 4  // 2 running + 2 new
+    1 * mesosDriver.launchTasks(_ as Collection<Protos.OfferID>, _ as List<Protos.TaskInfo>) // for place holder tasks
+    zeroNM.getTotalCapability().getVirtualCores() == 4 // capacity equals size of running + new containers
+    zeroNM.getTotalCapability().getMemory() == 4096
+    nodeStore.getNode(host).getContainerSnapshot() == null // container snapshot is released
+  }
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy:86]: more groovy:

```
    nodeStore.getNode(host).node.runningContainers.size() == 4  // 2 running + 2 new
    1 * mesosDriver.launchTasks(_ as Collection<Protos.OfferID>, _ as List<Protos.TaskInfo>) // for place holder tasks
    zeroNM.totalCapability.virtualCores == 4 // capacity equals size of running + new containers
    zeroNM.totalCapability.memory == 4096
    nodeStore.getNode(host).getContainerSnapshot() == null // container snapshot is released
```

File: myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+    def yarnNodeCapacityMgr = getYarnNodeCapacityManager()
+
+    when:
+    yarnNodeCapacityMgr.setNodeCapacity(zeroNM, Resources.createResource(2048, 2))
+
+    then:
+    zeroNM.getTotalCapability().getMemory() == 2048
+    zeroNM.getTotalCapability().getVirtualCores() == 2
+    1 * rmContext.getDispatcher().getEventHandler().handle(_ as NodeResourceUpdateSchedulerEvent)
+  }
```
Comment (kensipe) [inline myriad-scheduler/src/test/java/com/ebay/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy:100]: more groovy:

```
    zeroNM.totalCapability.memory == 2048
    zeroNM.totalCapability.virtualCores == 2
    1 * rmContext.dispatcher.eventHandler.handle(_ as NodeResourceUpdateSchedulerEvent)
```

**Comment by kensipe** [comment]
comments complete on the groovy spec review... IMO... it is fine without fixing most of what I commented on.  They are tests and work.   perhaps writing more java style is useful for people to understand what is going on.   After a detailed review, most of my comments are based on the groovy way and can be ignored.   **Nice work!!!**

**Comment by smarella** [comment]
@kensipe, appreciate your review of the PR. While I'd have loved to put your suggestions to use in this PR itself, I feel it's important to merge this in time for the 0.1 release.  I created https://issues.apache.org/jira/browse/MYRIAD-149 to improve the test code in the future.

**Comment by smarella** [comment]
Merged to apache/incubator-myriad.
https://github.com/apache/incubator-myriad/commit/645d17d8ad9263e59c18dea7110381b84643bc56

**Pull request closed without merging**